pkg/resourcetracker/app.go

```diff
@@ -0,0 +1,129 @@
+	return createResourceTracker(ctx, cli, app, getCurrentResourceTrackerName(app), v1beta1.ResourceTrackerTypeVersioned)
+}
+
+// CreateComponentRevisionResourceTracker create resourcetracker to record all component revision for application
+func CreateComponentRevisionResourceTracker(ctx context.Context, cli client.Client, app *v1beta1.Application) (*v1beta1.ResourceTracker, error) {
+	return createResourceTracker(ctx, cli, app, getComponentRevisionResourceTrackerName(app), v1beta1.ResourceTrackerTypeComponentRevision)
+}
+
+// ListApplicationResourceTrackers list resource trackers for application with all historyRTs sorted by version number
+func ListApplicationResourceTrackers(ctx context.Context, cli client.Client, app *v1beta1.Application) (rootRT *v1beta1.ResourceTracker, currentRT *v1beta1.ResourceTracker, historyRTs []*v1beta1.ResourceTracker, crRT *v1beta1.ResourceTracker, err error) {
+	rts := v1beta1.ResourceTrackerList{}
+	if err = cli.List(ctx, &rts, client.MatchingLabels{
+		oam.LabelAppName:      app.Name,
+		oam.LabelAppNamespace: app.Namespace,
+	}); err != nil {
+		return nil, nil, nil, nil, err
+	}
+	for _, rt := range rts.Items {
+		if rt.GetLabels() != nil && rt.GetLabels()[oam.LabelAppUID] != "" && rt.GetLabels()[oam.LabelAppUID] != string(app.UID) {
+			return nil, nil, nil, nil, fmt.Errorf("resourcetracker %s exists but controlled by another application (uid: %s), this could probably be cased by some mistakes while garbage collecting outdated resource. Please check this resourcetrakcer and delete it manually", rt.Name, rt.GetLabels()[oam.LabelAppUID])
+		}
+		switch rt.Spec.Type {
+		case v1beta1.ResourceTrackerTypeRoot:
+			rootRT = rt.DeepCopy()
+		case v1beta1.ResourceTrackerTypeVersioned:
+			if rt.Spec.ApplicationGeneration == app.GetGeneration() {
+				currentRT = rt.DeepCopy()
+			} else {
+				historyRTs = append(historyRTs, rt.DeepCopy())
+			}
+		case v1beta1.ResourceTrackerTypeComponentRevision:
+			crRT = rt.DeepCopy()
+		}
+	}
+	historyRTs = SortResourceTrackersByVersion(historyRTs, false)
+	return rootRT, currentRT, historyRTs, crRT, nil
+}
+
+// RecordManifestInResourceTracker records resources in ResourceTracker
+func RecordManifestInResourceTracker(ctx context.Context, cli client.Client, rt *v1beta1.ResourceTracker, manifest *unstructured.Unstructured, metaOnly bool) error {
+	if err := rt.AddManagedResource(manifest, metaOnly); err != nil {
+		return err
+	}
+	return cli.Update(ctx, rt)
+}
+
+// DeletedManifestInResourceTracker marks resources as deleted in resourcetracker, if remove is true, resources will be removed from resourcetracker
+func DeletedManifestInResourceTracker(ctx context.Context, cli client.Client, rt *v1beta1.ResourceTracker, manifest *unstructured.Unstructured, remove bool) error {
+	rt.DeleteManagedResource(manifest, remove)
+	return cli.Update(ctx, rt)
```

Comment:
where do you delete the ownerRef in resourceTracker for compatibility？

Reply:
The compatibility code has not been added yet. Will be added together with upgrade tests.